Include/internal/pycore_stackref.h

```diff
@@ -8,186 +8,150 @@ extern "C" {
 #  error "this header requires Py_BUILD_CORE define"
 #endif
 
+#include "pycore_object_deferred.h"
+
 #include <stddef.h>
 
 typedef union {
     uintptr_t bits;
 } _PyStackRef;
 
-static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 };
 
+#define Py_TAG_INT      (0)
 #define Py_TAG_DEFERRED (1)
+#define Py_TAG_RESERVED (2)
+#define Py_TAG_PTR      (3)
+
+#define Py_TAG          (Py_TAG_PTR)
+
+static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 | Py_TAG_DEFERRED};
+
+static inline int
+PyStackRef_IsNull(_PyStackRef stackref)
+{
+    return stackref.bits == Py_STACKREF_NULL.bits;
+}
+
+static inline int
+PyStackRef_IsDeferred(_PyStackRef ref)
+{
+    return ((ref.bits & Py_TAG) == Py_TAG_DEFERRED);
+}
 
 // Gets a PyObject * from a _PyStackRef
-#if defined(Py_GIL_DISABLED)
 static inline PyObject *
-PyStackRef_Get(_PyStackRef tagged)
+PyStackRef_AsPyObjectBorrow(_PyStackRef tagged)
 {
-    PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG_DEFERRED)));
+    PyObject *cleared = ((PyObject *)((tagged).bits & (~Py_TAG)));
     return cleared;
 }
-#else
-#   define PyStackRef_Get(tagged) ((PyObject *)((tagged).bits))
-#endif
 
-// Converts a PyObject * to a PyStackRef, stealing the reference.
-#if defined(Py_GIL_DISABLED)
+// Converts a PyObject * to a PyStackRef, stealing the reference
 static inline _PyStackRef
-_PyStackRef_StealRef(PyObject *obj)
+_PyStackRef_FromPyObjectSteal(PyObject *obj)
 {
     // Make sure we don't take an already tagged value.
-    assert(((uintptr_t)obj & Py_TAG_DEFERRED) == 0);
-    return ((_PyStackRef){.bits = ((uintptr_t)(obj))});
+    assert(((uintptr_t)obj & Py_TAG) == 0);
+    int tag = (obj == NULL || _Py_IsImmortal(obj)) ? (Py_TAG_DEFERRED) : Py_TAG_PTR;
```

Comment:
Like `PyStackRef_FromPyObjectNew`, I suspect we'll want to push `NULL` handling to the caller, but we can do that later.